reactive/kotlinx-coroutines-rx2/src/RxScheduler.kt

```diff
@@ -4,16 +4,133 @@
+        scheduleDirect(block, 0, TimeUnit.MILLISECONDS)
+
+    override fun scheduleDirect(block: Runnable, delay: Long, unit: TimeUnit): Disposable {
+        if (!scope.isActive) return Disposables.disposed()
+        val newBlock = RxJavaPlugins.onSchedule(block)
+        return scope.launch {
+            if (delay > 0) {
+                delay(unit.toMillis(delay))
+            }
+            newBlock.run()
+        }.asDisposable()
+    }
+
+    override fun createWorker(): Worker =
+        DispatcherWorker(dispatcher, schedulerJob).also {
+            it.start()
+        }
+
+    override fun shutdown() {
+        scope.cancel()
+    }
+
+    private class DispatcherWorker(dispatcher: CoroutineDispatcher, parentJob: Job) : Worker() {
+
+        private val workerJob = SupervisorJob(parentJob)
+        private val workerScope = CoroutineScope(workerJob + dispatcher)
+        private val blockChannel = Channel<SchedulerChannelTask>(Channel.UNLIMITED)
+
+        fun start() {
+            workerScope.launch {
+                while (isActive) {
+                    val task = blockChannel.receive()
+                    task.execute()
+                }
+            }
+        }
+
+        override fun isDisposed(): Boolean = !workerScope.isActive
+
+        override fun schedule(block: Runnable): Disposable =
+            schedule(block, 0, TimeUnit.MILLISECONDS)
+
+        override fun schedule(block: Runnable, delay: Long, unit: TimeUnit): Disposable {
+            if (!workerScope.isActive) return Disposables.disposed()
+
+            val newBlock = RxJavaPlugins.onSchedule(block)
+
+            val taskJob = Job(workerJob)
+            val task = SchedulerChannelTask(newBlock, taskJob)
+
+            if (delay == 0L) {
+                blockChannel.offer(task)
+            } else {
+                // Use `taskJob` as the parent here so the delay will also get cancelled if the Disposable
+                // is disposed.
+                workerScope.launch(taskJob) {
+                    // Delay *before* enqueuing the task, so other tasks (e.g. via schedule without delay)
+                    // aren't blocked by the delay.
+                    delay(unit.toMillis(delay))
+                    // Once the task is ready to run, it still needs to be executed via the queue to comply
+                    // with the Scheduler contract of running all worker tasks in a non-overlapping manner.
+                    blockChannel.offer(task)
+                }
+            }
+
+            return taskJob.asDisposable()
+        }
+
+        override fun dispose() {
+            workerScope.cancel()
+        }
+    }
+}
+
+
+/**
+ * Represents a task to be queued sequentially on a [Channel] for a [Scheduler.Worker].
+ *
+ * Delayed tasks do not block [Channel] from processing other tasks
+ */
+private class SchedulerChannelTask(
+    private val block: Runnable,
+    private val job: Job
```

Comment:
You don't need a `private val` on this `job` parameter. It is already stored by the `JobDisposable`. Just make a `DisposableJob.job` a `protected` property.